controllers/horizontal_runner_autoscaler_webhook.go

```diff
@@ -476,95 +477,99 @@ func (autoscaler *HorizontalRunnerAutoscalerGitHubWebhook) getScaleUpTargetWithF
 		return nil, nil
 	}
 
-	// Search for organization runner HRAs in default runner group
-	if target, err := scaleTarget(owner); err != nil {
-		log.Error(err, "finding organizational runner", "organization", owner)
-		return nil, err
-	} else if target != nil {
-		log.Info("job scale up target is organizational runners", "organization", owner)
-		return target, nil
-	}
-
-	if enterprise != "" {
-		// Search for enterprise runner HRAs in default runner group
-		if target, err := scaleTarget(enterpriseKey(enterprise)); err != nil {
-			log.Error(err, "finding enterprise runner", "enterprise", enterprise)
-			return nil, err
-		} else if target != nil {
-			log.Info("scale up target is default enterprise runners", "enterprise", enterprise)
-			return target, nil
-		}
-	}
-
-	// At this point there were no default organization/enterprise runners available to use, try now
-	// searching in runner groups
-
-	// We need to get the potential runner groups first to avoid spending API queries needless. Once/if GitHub improves an
+	// Find the potential runner groups first to avoid spending API queries needless. Once/if GitHub improves an
 	// API to find related/linked runner groups from a specific repository this logic could be removed
-	availableEnterpriseGroups, availableOrganizationGroups, err := autoscaler.getPotentialGroupsFromHRAs(ctx, enterprise, owner)
+	managedRunnerGroups, err := autoscaler.getManagedRunnerGroupsFromHRAs(ctx, enterprise, owner)
```

Comment:
I was carefully inspecting the implementation of `getManagedRunnerGroupsFromHRAs` and noticed that this function conflates runner groups associated with different enterprises and organizations, which means this works well only when you have only one organization and enterprise, but not when you have many of those.

The problem seems to stem from the fact that the returned data structure, `utils. RunnerGroups` struct, is unable to differentiate runner groups across enterprises and organizations.

This `managedRunnerGroups` is passed to the `GetRunnerGroupsVisibleToRepository` function, so it results in `GetRunnerGroupsVisibleToRepository` thinking an unmanaged runner group as if it is managed, which would result in ARC adding a runner (and the runner pod that hosts it) that is never used by GitHub Actions.

Probably we need to introduce namespacing to the data structure of `managedRunnerGroups`, either at the organization level or the enterprise level, so that ARC won't add a runner that is never used by GitHub.

I'll try to experiment a bit more so that I can hopefully come up with a better approach/implementation, but if you've any insights to share on it I'd appreciate it! Thanks.

Reply:
I found that, in getManagedRunnerGroupsFromHRAs, you did filter out any managed RunnerDeployments and RunnerSets that don't match the enterprise and the organization names read from the webhook event.
So, NVM! Probably it's not an issue.